package/samplers/auto_sampler/sampler.py

```diff
@@ -0,0 +1,243 @@
+from optuna.samplers import BaseSampler
+from optuna.samplers import CmaEsSampler
+from optuna.samplers import GPSampler
+from optuna.samplers import NSGAIIISampler
+from optuna.samplers import NSGAIISampler
+from optuna.samplers import RandomSampler
+from optuna.samplers import TPESampler
+from optuna.samplers._base import _process_constraints_after_trial
+from optuna.samplers._lazy_random_state import LazyRandomState
+from optuna.samplers._nsgaiii._sampler import _GENERATION_KEY as NSGA3_GENERATION_KEY
+from optuna.samplers.nsgaii._sampler import _GENERATION_KEY as NSGA2_GENERATION_KEY
+from optuna.search_space import IntersectionSearchSpace
+from optuna.trial import TrialState
+
+
+if TYPE_CHECKING:
+    from optuna.distributions import BaseDistribution
+    from optuna.study import Study
+    from optuna.trial import FrozenTrial
+
+
+MAXINT32 = (1 << 31) - 1
+THRESHOLD_OF_MANY_OBJECTIVES = 4
+
+
+class AutoSampler(BaseSampler):
+    _N_COMPLETE_TRIALS_FOR_CMAES = 250
+    _N_COMPLETE_TRIALS_FOR_NSGA = 1000
+
+    """Sampler automatically choosing an appropriate sampler based on search space.
+
+    This sampler is convenient when you are unsure what sampler to use.
+
+    Example:
+
+        .. testcode::
+
+            import optuna
+            import optunahub
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -5, 5)
+                return x**2
+
+            module = optunahub.load_module("samplers/auto_sampler")
+            study = optuna.create_study(sampler=module.AutoSampler())
+            study.optimize(objective, n_trials=300)
+
+    .. note::
+        This sampler requires optional dependencies of Optuna.
+        You can install them with ``pip install optunahub scipy torch cmaes``.
+        Alternatively, you can install them with ``pip install -r https://hub.optuna.org/samplers/auto_sampler/requirements.txt``.
+
+    Args:
+        seed: Seed for random number generator.
+        constraints_func:
+            An optional function that computes the objective constraints. It must take a
+            :class:`~optuna.trial.FrozenTrial` and return the constraints. The return value must
+            be a sequence of :obj:`float` s. A value strictly larger than 0 means that a
+            constraints is violated. A value equal to or smaller than 0 is considered feasible.
+            If ``constraints_func`` returns more than one value for a trial, that trial is
+            considered feasible if and only if all values are equal to 0 or smaller.
+
+            The ``constraints_func`` will be evaluated after each successful trial.
+            The function won't be called when trials fail or they are pruned, but this behavior is
+            subject to change in the future releases.
+    """
+
+    def __init__(
+        self,
+        seed: int | None = None,
+        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
```

Comment:
How about making all arguments keyword-only? It makes it easier to change the arguments in the future.

```suggestion
        self,
        *,
        seed: int | None = None,
        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
```